synapse/federation/sender/__init__.py

```diff
@@ -355,6 +355,39 @@ async def handle_event(event: EventBase) -> None:
                     if not is_mine and send_on_behalf_of is None:
                         return
 
+                    # We also want to not send out-of-band membership events.
+                    #
+                    # OOB memberships are used in three (and a half) situations:
+                    #
+                    # (1) invite events which we have received over federation. Those
+                    #     will have a `sender` on a different server, so will be
+                    #     skipped by the "is_mine" test above anyway.
+                    #
+                    # (2) rejections of invites to federated rooms. These are normally
+                    #     created via federation (in which case the remote server is
+                    #     responsible for sending out the rejection). If that fails,
+                    #     we'll create a leave event locally, but that's only really
+                    #     for the benefit of the invited user - we don't have enough
+                    #     information to send it out over federation.
+                    #
+                    # (2a) rescinded knocks. These are identical to rejected invites.
+                    #
+                    # (3) knock events which we have sent over federation. As with
+                    #     invite rejections, the remote server should send them out to
+                    #     the federation.
```

Comment:
Because we might not know all the servers joined to the room if we haven't joined it yet?

Reply:
Mostly because the spec says so, though now that I look at it the wording isn't that clear (https://spec.matrix.org/v1.2/server-server-api/#knocking-upon-a-room: "much like [joining rooms](https://spec.matrix.org/v1.2/server-server-api/#joining-rooms) the server must handshake their way into **having the knock sent on its behalf**".

Why does it say that? Principally by analogy to joining rooms, but that begs the question: why does it say it for joining rooms? I can't really remember. There's probably a good reason.

Reply:
Thanks. Happy to assume there's a good reason for now.

Reply:
According to Erik, the good reason is 

> The idea is it prevents a server from joining the room without telling other servers

(`s/join/knock/`)

Reply:
Though this is presumably a best-effort; we're trusting that the resident homeserver is going to push out our join/knock to everyone. But we already trusted the resident homeserver when we started the handshake anyway.